sensor: add support for reading euler angles with sensor api

tobias-mierzwa opened the pull request:
Add the possibility to read euler angels with sensor api.
For example the bno055 sensor from bosch supports euler angles. The
sensor is not yet supported from zephyr.

Signed-off-by: Tobias Mierzwa <tobias.mierzwa@mni.thm.de>

nashif commented on include/drivers/sensor.h:
```diff
@@ -80,7 +80,16 @@ enum sensor_channel {
-	SENSOR_CHAN_PRESS,
+        SENSOR_CHAN_PRESS,
+        /** Euler heading(yawn) in degrees */
+        SENSOR_CHAN_H,
```
You need to be more description in the name, I would use the full name here instead, maybe with Euler in the name as well. SENSOR_CHAN_H etc can mean lots of things...

Verfuchst replied:
Done, sorry, i was a bit fast with the commit :-)

avisconti commented on include/drivers/sensor.h:
```diff
@@ -80,7 +80,16 @@ enum sensor_channel {
 	SENSOR_CHAN_AMBIENT_TEMP,
 	/** Pressure in kilopascal. */
-	SENSOR_CHAN_PRESS,
+        SENSOR_CHAN_PRESS,
```
use <tab> and not <space> everywhere.
Infact, SENSOR_CHA_PRESS shouldn't be in the commit.

avisconti commented on include/drivers/sensor.h:
```diff
@@ -80,7 +80,16 @@ enum sensor_channel {
-	SENSOR_CHAN_PRESS,
+        SENSOR_CHAN_PRESS,
+        /** Euler heading(yawn) in degrees */
+        SENSOR_CHAN_EULER_H,
```
I would prefer SENSOR_CHAN_EULER_Y
Moreover, it is "yaw" not "yawn".

avisconti replied:
You can leave SENSOR_CHAN_EULER_H, but correct "yawn" into "yaw" in comment

avisconti replied:
Are you still using [space] here?

Verfuchst replied:
sorry now its tab, i messed that up in vi

avisconti commented on include/drivers/sensor.h:
```diff
@@ -80,7 +80,16 @@ enum sensor_channel {
+        /** Euler pitch in degrees */
+        SENSOR_CHAN_EULER_P,
+        /** Euler heading, roll and pitch in degrees */
+        SENSOR_CHAN_EULER_HRP,
```
YRP?
Not sure "yaw" is better than "heading".

Verfuchst replied:
I choosed heading because bosch also named it like that in the [datasheet](https://www.bosch-sensortec.com/media/boschsensortec/downloads/datasheets/bst-bno055-ds000.pdf) and in the [implementation](https://github.com/BoschSensortec/BNO055_driver/blob/master/bno055.h) too.

avisconti replied:
ok then

avisconti replied:
Pls, correct the other points

Verfuchst replied:
done

tobias-mierzwa committed:
sensor: add support for reading euler angles with sensor api

Add the possibility to read euler angels with sensor api.
For example the bno055 sensor from bosch supports euler angles. The
sensor is not yet supported from zephyr.

Signed-off-by: Tobias Mierzwa <tobias.mierzwa@web.de>

galak commented:
Is there something that will use this?

Verfuchst commented:
> Is there something that will use this?

None of the current sensors in Zephyr. I use this implementation for my bno055 driver implemenation in my own Zephyr repo. Next step is pull-request the bno055 sensor driver implemenation.

gmarull commented on include/drivers/sensor.h:
```diff
@@ -81,6 +81,15 @@ enum sensor_channel {
 	SENSOR_CHAN_AMBIENT_TEMP,
 	/** Pressure in kilopascal. */
 	SENSOR_CHAN_PRESS,
+	/** Euler heading(yaw) in degrees */
+	SENSOR_CHAN_EULER_H,
+	/** Euler roll in degrees */
+	SENSOR_CHAN_EULER_R,
+	/** Euler pitch in degrees */
+	SENSOR_CHAN_EULER_P,
+	/** Euler heading, roll and pitch in degrees */
+	SENSOR_CHAN_EULER_HRP,
```
I'd change the definitions to:

```suggestion
	/** Yaw/Heading angle in degrees */
	SENSOR_CHAN_YAW,
	/** Pitch angle in degrees */
	SENSOR_CHAN_PITCH,
	/** Roll angle in degrees */
	SENSOR_CHAN_ROLL,
	/** Yaw, pitch and roll andles in degrees */
	SENSOR_CHAN_YPR,
```

As far as I know, yaw, pitch and roll are always given as Euler angles.

MaureenHelm commented:
> > Is there something that will use this?
> 
> None of the current sensors in Zephyr. I use this implementation for my bno055 driver implemenation in my own Zephyr repo. Next step is pull-request the bno055 sensor driver implemenation.

I'd like to see the driver added as another commit in this pull request.

I'm concerned about adding new angular sensor channels with units in degrees, since existing angular sensor channels (`SENSOR_CHAN_GYRO_*`) use radians.

Verfuchst commented:
> > > Is there something that will use this?
> > 
> > 
> > None of the current sensors in Zephyr. I use this implementation for my bno055 driver implemenation in my own Zephyr repo. Next step is pull-request the bno055 sensor driver implemenation.
> 
> I'd like to see the driver added as another commit in this pull request.
> 
> I'm concerned about adding new angular sensor channels with units in degrees, since existing angular sensor channels (`SENSOR_CHAN_GYRO_*`) use radians.

Is a support of all functions from bno055 needed before merge? If this required, i would make it first as a standalone project till everything is supported.

MaureenHelm commented:
> Is the support of all functions of bno055 sensor needed before merge? If its required, i would first make it as a standalone project till everything is supported.

No, you can start by implementing basic functionality in the `sample_fetch` and `channel_get` APIs, then follow up with additional features in a future PR.

Verfuchst commented:
Hey, I just want to say that I have not forgotten the PR. I am just busy with my thesis atm and as soon as i am done with it, i will push the first version of bno055 driver :-)
